ldms/src/sampler/lustre_client/lustre_client_general.c

```diff
@@ -135,6 +154,7 @@ ldms_set_t llite_general_create(const char *producer_name,
         ldms_metric_array_set_str(set, index, fs_name);
         index = ldms_metric_by_name(set, "llite");
         ldms_metric_array_set_str(set, index, llite_name);
+	comp_id_helper_metric_update(set, cid);
```

Comment:
It looks to me like you are giving every filesystem the same component id. I suspect that people who care about component ids will want them to be different. This may also implies that your configuration handling function need to take an array or something. Keep in mind though that his plugin is fully dynamic, and we don't necessarily know which filesystem will appear in which order. So probably you need some way to specify a mapping between filesystem names and component IDs in the configuration. Technically a filesystem can be mounted twice and then you would have to somehow figure out which mount gets which component ID. Or may just accept that they are going to have the same component ID. But we at least need to handle the unique multiple fileystems use case, because that it a normal operating mode for this plugin.

Reply:
The network and file systems components (particularly in the case of this sampler) carry their own device identifier metrics. Component id as practiced is an enumerable (range() is well-defined) alias of hostnames (which being strings don't have a range()) for easy analysis. This allows collectives computation over related hosts. 

As with jobids, there is probably some ideal world in which compids could have been defined as arbitrarily long dotted string notation or both simply belong in a separate data set with a join performed at the database. Years back we chose to go the integer route for the analysis optimization reason.

Reply:
> The network and file systems components (particularly in the case of this sampler) carry their own device identifier metrics. Component id as practiced is an enumerable (range() is well-defined) alias of hostnames (which being strings don't have a range()) for easy analysis. This allows collectives computation over related hosts.
> 
> As with jobids, there is probably some ideal world in which compids could have been defined as arbitrarily long dotted string notation or both simply belong in a separate data set with a join performed at the database. Years back we chose to go the integer route for the analysis optimization reason.

That sounds awfully arbitrary and inconsistent to me. What metric sets that are expected to have multiple instances on a node do _not_ contain their own device identifier metrics? gpus contain identifier metrics. Does that mean that every gpu inside of a node gets the same component ID? CPUs are numbered. Does that mean they use the same component ID when they are in the same node?

Reply:
It's awfully arbitrary.  :-) It isn't inconsistent. The 'problem' is just that instead of being named component_id in the wayback (which it was when we didn't quite know what to do with it) the evolution of practice has been to something that might be reasonably called 'host_id'.  Everything inside the node gets the same component_id, just like everything inside the cluster gets the same cluster name in a slurm database.  The fact that the 'component_id' maps 1:1 with hostname is an artifact of producer being added to our schemas years after component_id practice was established.  

There was a prototype used of component_ids with hierarchical structures embedded in the single u64, where some bits were the host, some bits were a core number or gpu number or network device number or dimm number or what have you, but it was all pretty complicated and you had to play mask games and noone but the inventor understood it. Makes a lot more sense to just have device name metrics in device-type specific schema.

At this point, the weight of downstream tools assuming compid maps to hostname will, I expect, keep the usage the way it is.

Reply:
If you want to really pick at a legacy item (that doubles our data size), take a look at udata, which to my knowledge is no longer used for anything except making store code complicated. 
I would really like to remove *that*.

Reply:
Ah, that didn't sound like how they described it in the last call. But if that is the case, then so be it. But I think it would be good to add a comment to the man page under the component id option that explicitly states "all metric sets in this producer will use this same component id" or something to that effect.

Reply:
Since the user has no way of specifying alternate values per set instance, isn't that a given? But perhaps a bit of explanation of component_id added to the sampler_base page wouldn't be out of order.

Reply:
I am struggling with why adding a single uint64_t metric to a set requires touching 11 files, adding two new files and introducing a handful of new API.

If we want to hold court on how component_id should be managed, then let's do that in an issue or on the wiki, but not shoe horn API design issues into a change that would otherwise require about 3 lines of code.

Reply:
No, I don't think it necessarily is. The plugin might try to disambiguate them in some way, like it does for instances. I've had repeated lack of clarity over comp ID through the past year, so explicit comments like I requested would be very helpful.

Reply:
added compid explanation to man page.